### PR TITLE
feat(angular slider)

### PR DESCRIPTION
**Pull request by skytracer**
An angular slider based on progress-bar and button

Plunker here
http://plnkr.co/edit/nHJHKGlMiPHzmjk6fO3t?p=preview

### CODE REVIEW

File: src/slider/docs/demo.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+var SliderDemoCtrl = function($scope) {
```
Comment (wesleycho) [inline src/slider/docs/demo.js]: Add a line for `angular.module('ui.bootstrap.slider').controller('SliderDemoCtrl', SliderDemoCtrl);`

File: src/slider/docs/demo.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        <label for="sliderelem3">update on drag</label>
+        <slider id="sliderelem3" update-on-drag="true" transition="1s ease 0s" ng-model="dynamic">{{dynamic}}</slider>
+    </div>
+    <div class="form-group">
+        <label for="sliderelem4">no text in button</label>
+        <slider id="sliderelem4" update-on-drag="true" transition="1s ease 0s" ng-model="dynamic"></slider>
+    </div>
+    <div class="form-group">
+        <label for="sliderelem4">add &amp;nbsp text for handle height as having text</label>
+        <slider id="sliderelem4"  ng-model="dynamic">&nbsp</slider>
```
Comment (wesleycho) [inline src/slider/docs/demo.html]: Remove extra space before ng-model

File: src/slider/slider.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        scope.transition = scope.transition || sliderConfig['transition'];
+        scope.updateOnDrag = angular.isDefined(attrs.updateOnDrag) ? scope.$parent.$eval(attrs.updateOnDrag) : (sliderConfig['updateOnDrag'] || false);
+
+        scope.btn = element.find('a');
+        scope.bar = element.children().eq(1).children().eq(0);
+        scope.range = scope.max - scope.min;
+
+        scope.hasAnimation = (scope.transition) && (scope.transition !== 'none');
+
+        scope.setAnimation = function(doanimate) {
```
Comment (wesleycho) [inline src/slider/slider.js]: I would recommend moving this function into a directive controller and requiring that controller.

File: src/slider/slider.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          {
+            scope.btn.css({'transition-property': 'left'});
+            scope.bar.css({'transition-property': 'width'});
+          } else {
+            scope.btn.css({'transition-property': 'none'});
+            scope.bar.css({'transition-property': 'none'});
+          }
+        };
+
+        scope.init = function() {
```
Comment (wesleycho) [inline src/slider/slider.js]: Same with this init function

File: src/slider/slider.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+          if(scope.hasAnimation)
+          {
+            scope.btn.css({'transition': 'left ' + scope.transition});
+            scope.bar.css({'transition': 'width ' + scope.transition});
+          }
+          scope.setAnimation(scope.hasAnimation);
+        };
+
+        scope.getPercentage = function(value) {
+          return angular.isNumber(value) ? Math.max(0, Math.min(100, Math.round(100 * (value - scope.min) / scope.range))) : 0;
```
Comment (wesleycho) [inline src/slider/slider.js]: This does not adequately screen out `NaN`, which is a number.

File: src/slider/slider.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        scope.setValid = function(value) {
+          var v = parseInt(value, 10);
+          ngModel.$setValidity('slider', (scope.max >= v) && (v >= scope.min));
+          return v;
+        };
+
+        ngModel.$formatters.push(scope.setValid);
+        ngModel.$parsers.push(scope.setValid);
+
+        scope.update = function() {
```
Comment (wesleycho) [inline src/slider/slider.js]: This function should really be `ngModel.$render`

File: template/slider/slider.html [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<div class="progress-slider" style="position: relative;margin-top: 15px;">
+  <a role=["buton"] ng-class="btnClass || 'btn-default btn-lg'" class="btn" style="margin-left:-15px;display:inline-block;position: relative;z-index: 1" ng-transclude></a>
```
Comment (wesleycho) [inline template/slider/slider.html]: Typo - should be `role="button"`

File: src/slider/slider.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+          ngModel.$setViewValue(scope.percentageToValue(scope.percent));
+          scope.$apply();
+        };
+
+        ngModel.$render = function() {
+          scope.percent = scope.getPercentage(ngModel.$viewValue);
```
Comment (wesleycho) [inline src/slider/slider.js]: If `ngModel.$setViewValue` is called, the $parsers are run and the model value gets updated - what should be done is the `setValid` parser should get pushed into `ngModel.$parsers` and return the percentage value.

File: src/slider/slider.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            scope.bar.css({'transition': 'width ' + scope.transition});
+          }
+          scope.setAnimation(scope.hasAnimation);
+        };
+
+        scope.getPercentage = function(value) {
+          return angular.isNumber(value) ? Math.max(0, Math.min(100, Math.round(100 * (value - scope.min) / scope.range))) : 0;
+        };
+
+        scope.percentageToValue = function(perc) {
```
Comment (wesleycho) [inline src/slider/slider.js]: Remove from being on `scope`

File: src/slider/slider.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+          return Math.round(scope.min + (scope.range * perc / 100));
+        };
+
+        scope.setValid = function(value) {
```
Comment (wesleycho) [inline src/slider/slider.js]: Make this an anonymous function also.

**Comment by rvanbaalen** [comment]
@skytracer Any updates on this?

**Comment by wesleycho** [comment]
Thank you for your effort here.

This is a cool widget, but at this point, I believe this is best open sourced separately - we are currently underwater with issues with existing components, and just started on migration to Angular 2 on top of that. There are issues with git history here that need to be resolved, as well as some refactoring that needs to be done to properly use ngModel.

Closing in consideration of these points.

**Pull request closed without merging**